pkg/cloud/services/ec2/instances.go

```diff
@@ -222,13 +222,21 @@ func (s *Service) CreateInstance(scope *scope.MachineScope, userData []byte) (*i
 
 	// If SSHKeyName WAS NOT provided in the AWSMachine Spec, fallback to the value provided in the AWSCluster Spec.
 	// If a value was not provided in the AWSCluster Spec, then use the defaultSSHKeyName
-	input.SSHKeyName = scope.AWSMachine.Spec.SSHKeyName
-	if input.SSHKeyName == nil {
-		if scope.AWSCluster.Spec.SSHKeyName != nil {
+	// Note that:
+	// - a nil AWSMachine.Spec.SSHKeyName value means use the AWSCluster.Spec.SSHKeyName SSH key name value
+	// - nil values for both AWSCluster.Spec.SSHKeyName and AWSMachine.Spec.SSHKeyName means use the default SSH key name value
+	// - an empty string means do not set an SSH key name at all
+	// - otherwise use the value specified in either AWSMachine or AWSCluster
+	if scope.AWSMachine.Spec.SSHKeyName != nil {
+		if *scope.AWSMachine.Spec.SSHKeyName != "" {
+			input.SSHKeyName = scope.AWSMachine.Spec.SSHKeyName
+		}
+	} else if scope.AWSCluster.Spec.SSHKeyName != nil {
+		if *scope.AWSCluster.Spec.SSHKeyName != "" {
 			input.SSHKeyName = scope.AWSCluster.Spec.SSHKeyName
```

Comment:
I think it's possible to fall through this `if` without setting a value for `input.SSHKeyName` – is that intentional?

If not, consider:
```suggestion
	if scope.AWSMachine.Spec.SSHKeyName != nil && *scope.AWSMachine.Spec.SSHKeyName != "" {
		input.SSHKeyName = scope.AWSMachine.Spec.SSHKeyName
	} else if scope.AWSCluster.Spec.SSHKeyName != nil && *scope.AWSCluster.Spec.SSHKeyName != "" {
		input.SSHKeyName = scope.AWSCluster.Spec.SSHKeyName
```

Reply:
Oh, now I see it: 

>  an empty string means do not set an SSH key name at all

My mistake, sorry. I believe the code implements the spec in the comment, but it might be worth an issue around cleaning up this list of options + fallbacks for v1alpha4?

Reply:
I'm wondering if we can simplify this logic a bit with something like this instead:
```go
	switch {
	case scope.AWSMachine.Spec.SSHKeyName != nil:
		// prefer AWSMachine.Spec.SSHKeyName if it is defined
		input.SSHKeyName = scope.AWSMachine.Spec.SSHKeyName
	case scope.AWSCluster.Spec.SSHKeyName != nil:
		// fallback to AWSCluster.Spec.SSHKeyName if it is defined
		input.SSHKeyName = scope.AWSCluster.Spec.SSHKeyName
	default:
		input.SSHKeyName = aws.String(defaultSSHKeyName)
	}
```

As far as I can tell, there really isn't any reason to treat an empty string as any different than an explicit value as far as the fallback logic and using a switch potentially helps remove some of the complexity related to trying to grok the nested conditionals.

Reply:
> As far as I can tell, there really isn't any reason to treat an empty string as any different than an explicit value as far as the fallback logic and using a switch potentially helps remove some of the complexity related to trying to grok the nested conditionals.

There is a meaningful distinction, which is not setting a key name vs setting the default key name. Maybe adding a `dontUseSSHKeys` boolean would simplify things?

Reply:
> There is a meaningful distinction, which is not setting a key name vs setting the default key name. Maybe adding a dontUseSSHKeys boolean would simplify things?

Ah, I think I'm following now. Apologies about the previous confusion.

What if we separated the fallback handling from the handling of empty string? I think it might help clear up some of the confusion around trying to handle both fallback logic and the impedance mismatch between our api and the aws api a bit better.

```go
	// handle fallback prioritization
	var prioritizedSSHKeyName string
	switch {
	case scope.AWSMachine.Spec.SSHKeyName != nil:
		// prefer AWSMachine.Spec.SSHKeyName if it is defined
		prioritizedSSHKeyName = *scope.AWSMachine.Spec.SSHKeyName
	case scope.AWSCluster.Spec.SSHKeyName != nil:
		// fallback to AWSCluster.Spec.SSHKeyName if it is defined
		prioritizedSSHKeyName = *scope.AWSCluster.Spec.SSHKeyName
	default:
		prioritizedSSHKeyName = defaultSSHKeyName
	}

	// Only set input.SSHKeyName if the user did not explicitly request no ssh key be set (explicitly setting "" on either the Machine or related Cluster)
	if prioritizedSSHKeyName != "" {
		input.SSHKeyName = aws.String(prioritiziedSSHKeyName)
	}
```

Reply:
Is it possible to not have an ssh key at all? I would prefer to not use ssh and instead connect with session manager but it doesn't appear possible because AWS_SSH_KEY_NAME is required.

Reply:
@rothgar that is a good point, the template handling through clusterctl does seem to make consuming this a bit difficult today.

Reply:
The idea is that if one does not specify a key to use then we would use the configured default, but if one explicitly specifies no key with an empty string, then we would explicitly not use a key when creating the instance.

Reply:
I'm fine with `AWS_SSH_KEY_NAME=""` hard to make it explicit while also not letting someone lock themselves out.

Reply:
If ssh is explicitly set to "" there should also be logic to not create the basiton security group. Are there other resources created to allow SSH?

Reply:
> If ssh is explicitly set to "" there should also be logic to not create the basiton security group. Are there other resources created to allow SSH?

I believe longer term we are planning on allowing more customization here. It gets a bit tricky because this value can be overridden through a similar setting on Machine* resources, so we can't necessarily assume that if it is set to "" on the Cluster that it isn't set to a different value on one or more Machines that belong to the Cluster.